### PR TITLE
Feat/extcodehash

### PR DESCRIPTION
**Pull request by Eikix**
## Pull Request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

Resolves: #131

### CODE REVIEW

File: crates/contracts/src/tests/test_kakarot_core.cairo [hunk 1]
```diff
@@ -82,6 +82,20 @@ fn test_kakarot_core_deploy_eoa() {
     assert(event.starknet_address == eoa_starknet_address, 'wrong starknet address');
 }
 
+#[test]
+#[available_gas(20000000)]
```
Comment (Eikix) [inline crates/contracts/src/tests/test_kakarot_core.cairo:86]: added this missing test

File: crates/contracts/src/tests/utils.cairo [hunk 2]
```diff
@@ -94,3 +94,309 @@ fn fund_account_with_native_token(
     testing::set_contract_address(constants::ETH_BANK());
     native_token.transfer(contract_address, amount);
 }
+
+// Counter Smart Contract Bytecode:
```
Comment (Eikix) [inline crates/contracts/src/tests/utils.cairo:98]: In Kakarot cairo0, we have foundry as a dependency and have integration solidity tests
here, I think this is sufficient
but we might want to add them as we move on

Comment (Eikix) [reply crates/contracts/src/tests/utils.cairo:98]: [reproduce link](https://emn178.github.io/online-tools/keccak_256.html?input=6080604052348015600f57600080fd5b506004361060465760003560e01c806306661abd14604b578063371303c01460655780636d4ce63c14606d578063b3bcfa82146074575b600080fd5b605360005481565b60405190815260200160405180910390f35b606b607a565b005b6000546053565b606b6091565b6001600080828254608a919060b7565b9091555050565b6001600080828254608a919060cd565b634e487b7160e01b600052601160045260246000fd5b8082018082111560c75760c760a1565b92915050565b8181038181111560c75760c760a156fea2646970667358221220f379b9089b70e8e00da8545f9a86f648441fdf27ece9ade2c71653b12fb80c7964736f6c63430008120033&input_type=hex)

File: crates/evm/src/tests/test_instructions/test_environment_information.cairo [hunk 3]
```diff
@@ -710,3 +714,108 @@ fn test_returndata_copy(dest_offset: u32, offset: u32, mut size: u32) {
...
+    machine.exec_extcodehash().unwrap();
+
+    // Then
+    machine.stack.peek().unwrap().print();
+    assert(
+        machine
+            .stack
+            .peek()
+            // extcodehash(Counter.sol) := 0x82abf19c13d2262cc530f54956af7e4ec1f45f637238ed35ed7400a3409fd275 (source: remix)
+            // https://emn178.github.io/online-tools/keccak_256.html?input=6080604052348015600f57600080fd5b506004361060465760003560e01c806306661abd14604b578063371303c01460655780636d4ce63c14606d578063b3bcfa82146074575b600080fd5b605360005481565b60405190815260200160405180910390f35b606b607a565b005b6000546053565b606b6091565b6001600080828254608a919060b7565b9091555050565b6001600080828254608a919060cd565b634e487b7160e01b600052601160045260246000fd5b8082018082111560c75760c760a1565b92915050565b8181038181111560c75760c760a156fea2646970667358221220f379b9089b70e8e00da8545f9a86f648441fdf27ece9ade2c71653b12fb80c7964736f6c63430008120033&input_type=hex
```
Comment (Eikix) [inline crates/evm/src/tests/test_instructions/test_environment_information.cairo:817]: [reproduce link](https://emn178.github.io/online-tools/keccak_256.html?input=6080604052348015600f57600080fd5b506004361060465760003560e01c806306661abd14604b578063371303c01460655780636d4ce63c14606d578063b3bcfa82146074575b600080fd5b605360005481565b60405190815260200160405180910390f35b606b607a565b005b6000546053565b606b6091565b6001600080828254608a919060b7565b9091555050565b6001600080828254608a919060cd565b634e487b7160e01b600052601160045260246000fd5b8082018082111560c75760c760a1565b92915050565b8181038181111560c75760c760a156fea2646970667358221220f379b9089b70e8e00da8545f9a86f648441fdf27ece9ade2c71653b12fb80c7964736f6c63430008120033&input_type=hex)

File: crates/contracts/src/tests/utils.cairo (hunk 2 above)
Comment (enitrat) [inline crates/contracts/src/tests/utils.cairo:102]: Could be in a `test_data` folder

File: crates/evm/src/instructions/environmental_information.cairo [hunk 4]
```diff
@@ -212,8 +217,39 @@ impl EnvironmentInformationImpl of EnvironmentInformationTrait {
...
+    // Else return, the hash of the account's code
     /// # Specification: https://www.evm.codes/#3f?fork=shanghai
     fn exec_extcodehash(ref self: Machine) -> Result<(), EVMError> {
-        Result::Ok(())
+        let evm_address = self.stack.pop_eth_address()?;
+
+        let maybe_account = AccountTrait::account_at(evm_address)?;
+        let account = match maybe_account {
+            Option::Some(account) => account,
+            // Case 1: The address corresponds to a non-existing or destroyed contract account.
```
Comment (enitrat) [inline crates/evm/src/instructions/environmental_information.cairo:231]: >  // Case 1: The address corresponds to a non-existing or destroyed contract account.

You already documented this

>     // If the account does not exist of was destroyed (SELFDESTRUCT), return 0

File: crates/evm/src/instructions/environmental_information.cairo [hunk 5]
```diff
@@ -212,8 +217,39 @@ impl EnvironmentInformationImpl of EnvironmentInformationTrait {
...
+            Option::None => { return self.stack.push(0); },
+        };
+
+        match account {
+            Account::EOA(eoa) => {
+                // Case 2: The address corresponds to a EOA. If so, the account exists but has no code.
```
Comment (enitrat) [inline crates/evm/src/instructions/environmental_information.cairo:237]: > // If the account has no code (or is a precompile), return the empty hash:
    // `0xc5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a470`
    
    You already documented this

File: crates/evm/src/tests/test_instructions/test_environment_information.cairo [hunk 6]
```diff
@@ -1,15 +1,19 @@
-use contracts::kakarot_core::interface::IExtendedKakarotCoreDispatcherTrait;
+use contracts::kakarot_core::{interface::IExtendedKakarotCoreDispatcherImpl, KakarotCore};
+use contracts::tests::utils::counter_evm_bytecode;
 use contracts::tests::utils::{
     deploy_kakarot_core, deploy_native_token, fund_account_with_native_token
 };
+use debug::U256PrintImpl;
```
Comment (enitrat) [inline crates/evm/src/tests/test_instructions/test_environment_information.cairo:6]:
```suggestion
```

**Comment by Eikix** [comment]
Alrighty!

**Review by enitrat — APPROVED** [review]
LGTM

**Pull request merged**